src/trace/implementations/merge_batcher_col.rs

```diff
@@ -0,0 +1,353 @@
+        match batch {
+            RefOrMut::Ref(reference) => {
+                // This is a moment at which we could capture the allocations backing
+                // `batch` into a different form of region, rather than just  cloning.
+                self.sorter.push(&mut reference.clone());
+            },
+            RefOrMut::Mut(reference) => {
+                self.sorter.push(reference);
+            }
+        }
+    }
+
+    // Sealing a batch means finding those updates with times not greater or equal to any time
+    // in `upper`. All updates must have time greater or equal to the previously used `upper`,
+    // which we call `lower`, by assumption that after sealing a batcher we receive no more
+    // updates with times not greater or equal to `upper`.
+    #[inline]
+    fn seal(&mut self, upper: Antichain<B::Time>) -> B {
+
+        let mut builder = B::Builder::new();
+
+        let mut merged = Default::default();
+        self.sorter.finish_into(&mut merged);
+
+        let mut kept = Vec::new();
+        let mut keep = TimelyStack::default();
+
+        self.frontier.clear();
+
+        for mut buffer in merged.drain(..) {
+            let mut last: Option<&((_, _), _, _)> = None;
+            for datum @ ((key, val), time, diff) in &buffer[..] {
+                if upper.less_equal(time) {
+                    self.frontier.insert(time.clone());
+                    if !keep.is_empty() && keep.len() == keep.capacity() {
+                        kept.push(keep);
+                        keep = self.sorter.empty();
+                    }
+                    keep.copy(datum);
+                }
+                else {
+                    builder.push((key.clone(), val.clone(), time.clone(), diff.clone()));
+                }
+                if let Some(last) = last {
+                    assert!(last <= datum);
+                }
+                last = Some(datum);
+            }
+            // Recycling buffer.
+            buffer.clear();
+            self.sorter.recycle(buffer);
+        }
+
+        // Finish the kept data.
+        if !keep.is_empty() {
+            kept.push(keep);
+        }
+        if !kept.is_empty() {
+            self.sorter.push_list(kept);
+        }
+
+        // Drain buffers (fast reclamation).
+        self.sorter.clear_stash();
+
+        let seal = builder.done(self.lower.clone(), upper.clone(), Antichain::from_elem(<B::Time as timely::progress::Timestamp>::minimum()));
+        self.lower = upper;
+        seal
+    }
+
+    // the frontier of elements remaining after the most recent call to `self.seal`.
+    fn frontier(&mut self) -> timely::progress::frontier::AntichainRef<B::Time> {
+        self.frontier.borrow()
+    }
+}
+
+struct TimelyStackQueue<T: Columnation> {
+    list: TimelyStack<T>,
+    head: usize,
+}
+
+impl<T: Columnation + 'static> TimelyStackQueue<T> {
+
+    fn new() -> Self { TimelyStackQueue::from(Default::default()) }
```

Comment:
We check on recycling whether the capacity is the default capacity and only then retain the allocation. Otherwise, we never write to a `TimelyStackQueue`.